File: backend/modules/production_saas_upgrade/agreement_engine.py
from __future__ import annotations

from typing import Any

from .document_generator import DocumentGenerator


class AgreementEngine:
    """Additive agreement engine for production SaaS module layer."""

    @staticmethod
    def execute(*, payload: dict[str, Any], actor_user_id: str, signature_method: str, signature_record: str, ip_address: str | None, device_info: str | None):
        template = "medical_equipment_agreement" if payload.get("agreement_type") == "medical_equipment" else "home_care_agreement"
        return DocumentGenerator.generate(
            template_key=template,  # type: ignore[arg-type]
            title="Medico-Legal Agreement",
            payload=payload,
            actor_user_id=actor_user_id,
            signature_method=signature_method,
            signature_record=signature_record,
            ip_address=ip_address,
            device_info=device_info,
        )

File: frontend/src/lib/server/platform/agreement-engine.ts
import type { Prisma } from "@prisma/client";
import { generateLegalDocument } from "@/lib/server/platform/document-automation";
import type { SignatureMethod } from "@/lib/server/platform/types";

export async function runAgreementEngine(input: {
  tenantId: string;
  actorUserId: string;
  caseId?: string;
  payload: Prisma.InputJsonValue;
  signatureMethod: SignatureMethod;
  signatureRecord: string;
  requestIp: string | null;
  deviceInfo: string | null;
}) {
  const subtype = (input.payload as Record<string, unknown> | null)?.agreement_type;
  const templateKey = subtype === "medical_equipment" ? "medical_equipment_agreement" : "home_care_agreement";

  return generateLegalDocument({
    tenantId: input.tenantId,
    actorUserId: input.actorUserId,
    caseId: input.caseId,
    payload: input.payload,
    templateKey,
    title: "Medico-Legal Agreement",
    titleAr: "اتفاقية طبية قانونية",
    signatureMethod: input.signatureMethod,
    signatureRecord: input.signatureRecord,
    requestIp: input.requestIp,
    deviceInfo: input.deviceInfo,
  });
}
